src/Microsoft.TemplateEngine.Core/Util/TokenTrie.cs

```diff
@@ -1,129 +1,89 @@
+﻿using System.Collections.Generic;
 using Microsoft.TemplateEngine.Core.Contracts;
+using Microsoft.TemplateEngine.Core.Matching;
 
 namespace Microsoft.TemplateEngine.Core.Util
 {
-    public class TokenTrie : ITokenTrie
+    public class TokenTrie : Trie<Token>, ITokenTrie
     {
-        private readonly Dictionary<byte, TokenTrie> _map = new Dictionary<byte, TokenTrie>();
-        private readonly List<int> _tokenLength;
-        private readonly List<byte[]> _tokens;
+        private List<IToken> _tokens = new List<IToken>();
+        private List<int> _lengths = new List<int>();
 
-        public TokenTrie()
-        {
-            Index = -1;
-            _tokenLength = new List<int>();
-            _tokens = new List<byte[]>();
-        }
-
-        public int Count { get; private set; }
-
-        public int Index { get; private set; }
+        public int Count => _tokens.Count;
 
         public int MaxLength { get; private set; }
 
-        public int MinLength { get; private set; }
+        public int MinLength { get; private set; } = int.MaxValue;
 
-        public IReadOnlyList<byte[]> Tokens => _tokens;
+        public IReadOnlyList<int> TokenLength => _lengths;
 
-        public IReadOnlyList<int> TokenLength => _tokenLength;
+        public IReadOnlyList<IToken> Tokens => _tokens;
 
-        public int AddToken(byte[] token)
+        public int AddToken(byte[] literalToken)
         {
-            int result = Count;
-            AddToken(token, result);
-            return result;
+            return AddToken(TokenConfig.LiteralToken(literalToken));
         }
 
-        public void AddToken(byte[] token, int index)
+        public void AddToken(byte[] literalToken, int index)
         {
-            ++Count;
-            TokenTrie current = this;
-            _tokenLength.Add(token.Length);
-            _tokens.Add(token);
-
-            MaxLength = Math.Max(MaxLength, token.Length);
-
-            MinLength = MinLength == 0
-                ? token.Length
-                : Math.Min(MinLength, token.Length);
+            AddToken(TokenConfig.LiteralToken(literalToken), index);
+        }
 
-            for (int i = 0; i < token.Length; ++i)
-            {
-                TokenTrie child;
-                if (!current._map.TryGetValue(token[i], out child))
-                {
-                    child = new TokenTrie();
-                    current._map[token[i]] = child;
-                }
-
-                if (i == token.Length - 1)
-                {
-                    child.Index = index;
-                }
-
-                current = child;
-            }
+        public int AddToken(IToken token)
+        {
+            int count = _tokens.Count;
+            AddToken(token, count);
+            return count;
         }
 
-        public bool GetOperation(byte[] buffer, int bufferLength, ref int currentBufferPosition, out int token)
+        public void AddToken(IToken token, int index)
         {
```

Comment:
I don't see any checking that multiple tokens don't have the same index. This would only potentially be a problem if tokens are added with explicit indexes.
OTOH, I don't have this fully internalized yet, so perhaps duplicated indexes are ok.